File: src/adapters/mcp-compiler.ts
import { execFileSync } from 'node:child_process'
import { homedir } from 'node:os'
import path from 'node:path'
import type { FileRecord, SetupScope, ToolId } from '../types.js'
import { ensureDir, fileExists, fileHash, readFile, writeFile } from '../utils/files.js'
import { stripJsonComments } from '../utils/jsonc.js'

interface McpServer {
  description?: string
  command?: string
  args?: string[]
  env?: Record<string, string>
  url?: string
  headers?: Record<string, string>
  tools?: string[]
  enabled?: boolean
}

interface McpCatalog {
  servers: Record<string, McpServer>
}

interface CopilotPromptInput {
  type: 'promptString'
  id: string
  description: string
  password: true
}

function readCanonicalMcp(targetDir: string): McpCatalog | null {
  const mcpPath = path.join(targetDir, '.ai', 'mcp.json')
  if (!fileExists(mcpPath)) return null
  try {
    return JSON.parse(readFile(mcpPath)) as McpCatalog
  } catch {
    return null
  }
}

function getEnabledServers(catalog: McpCatalog): Record<string, McpServer> {
  const result: Record<string, McpServer> = {}
  for (const [name, server] of Object.entries(catalog.servers)) {
    if (server.enabled !== false) {
      result[name] = server
    }
  }
  return result
}

function toMcpJson(servers: Record<string, McpServer>): Record<string, unknown> {
  return { mcpServers: toClaudeCodeMcpInner(servers) }
}

function toClaudeCodeMcpInner(servers: Record<string, McpServer>): Record<string, unknown> {
  const mcpServers: Record<string, unknown> = {}
  for (const [name, server] of Object.entries(servers)) {
    if (server.url) {
      const entry: Record<string, unknown> = { url: server.url }
      if (server.headers) entry.headers = server.headers
      mcpServers[name] = entry
      continue
    }

    const entry: Record<string, unknown> = {
      command: server.command,
      args: server.args,
    }
    if (server.env) entry.env = server.env
    mcpServers[name] = entry
  }
  return mcpServers
}

function toOpenCodeJsonc(allServers: Record<string, McpServer>): Record<string, unknown> {
  const mcp: Record<string, unknown> = {}
  for (const [name, server] of Object.entries(allServers)) {
    const isEnabled = server.enabled !== false
    if (server.url) {
      mcp[name] = {
        type: 'remote',
        enabled: isEnabled,
        url: server.url,
        ...(server.headers ? { headers: transformEnvSyntax(server.headers, '{env:$1}') } : {}),
      }
      continue
    }

    const entry: Record<string, unknown> = {
      type: 'local',
      enabled: isEnabled,
      command: [server.command, ...(server.args ?? [])],
    }
    if (server.env) entry.environment = transformEnvSyntax(server.env, '{env:$1}')
    mcp[name] = entry
  }
  return mcp
}

function mergeOpenCodeMcp(existingRaw: unknown, managed: Record<string, unknown>): Record<string, unknown> {
  const merged: Record<string, unknown> = {}

  if (existingRaw && typeof existingRaw === 'object' && !Array.isArray(existingRaw)) {
    for (const [name, entry] of Object.entries(existingRaw)) {
      if (!(name in managed)) {
        merged[name] = entry
      }
    }
  }

  for (const [name, entry] of Object.entries(managed)) {
    merged[name] = entry
  }

  return merged
}

function toCopilotServerEntries(servers: Record<string, McpServer>): {
  entries: Record<string, unknown>
  placeholderIds: Set<string>
} {
  const entries: Record<string, unknown> = {}
  const placeholderIds = new Set<string>()

  for (const [name, server] of Object.entries(servers)) {
    if (server.url) {
      const entry: Record<string, unknown> = {
        type: 'sse',
        url: server.url,
      }
      if (server.headers) entry.headers = server.headers
      entries[name] = entry
      continue
    }

    const entry: Record<string, unknown> = {
      type: 'stdio',
      command: server.command,
      args: server.args,
    }
    if (server.env) {
      entry.env = server.env
      for (const value of Object.values(server.env)) {
        for (const match of value.matchAll(/\$\{(\w+)\}/g)) {
          if (match[1]) placeholderIds.add(match[1])
        }
      }
    }
    entries[name] = entry
  }

  return { entries, placeholderIds }
}

function toCopilotPromptInputs(placeholderIds: Set<string>): CopilotPromptInput[] {
  return [...placeholderIds]
    .sort((a, b) => a.localeCompare(b))
    .map((id) => ({
      type: 'promptString',
      id,
      description: id,
      password: true,
    }))
}

function toCopilotVSCodeMcp(servers: Record<string, McpServer>): Record<string, unknown> {
  const { entries, placeholderIds } = toCopilotServerEntries(servers)
  const output: Record<string, unknown> = { servers: entries }
  if (placeholderIds.size > 0) {
    output.inputs = toCopilotPromptInputs(placeholderIds)
  }
  return output
}

function toCopilotCliMcp(servers: Record<string, McpServer>): Record<string, unknown> {
  const { entries } = toCopilotServerEntries(servers)
  return { mcpServers: entries }
}

function toGeminiSettings(servers: Record<string, McpServer>): Record<string, unknown> {
  const mcpServers: Record<string, unknown> = {}
  for (const [name, server] of Object.entries(servers)) {
    if (server.url) {
      console.warn(`⚠️  Skipping remote server "${name}" for gemini (not supported)`)
      continue
    }
    const entry: Record<string, unknown> = {
      command: server.command,
      args: server.args,
    }
    if (server.env) entry.env = transformEnvSyntax(server.env, '$$$1')
    mcpServers[name] = entry
  }
  return { mcpServers }
}

function transformEnvSyntax(envObj: Record<string, string>, targetPattern: string): Record<string, string> {
  const result: Record<string, string> = {}
  for (const [key, value] of Object.entries(envObj)) {
    result[key] = value.replace(/\$\{(\w+)\}/g, targetPattern)
  }
  return result
}

function resolveToolRoot(toolId: ToolId, targetDir: string, setupScope: SetupScope): string {
  if (setupScope === 'global') return targetDir

  switch (toolId) {
    case 'opencode':
      return path.join(targetDir, '.opencode')
    case 'claude-code':
      return path.join(targetDir, '.claude')
    case 'gemini':
      return path.join(targetDir, '.gemini')
    case 'codex':
      return path.join(targetDir, '.codex')
    default:
      return targetDir
  }
}

function resolveHomeDir(homeDir?: string): string {
  return homeDir ?? homedir()
}

function mergeJsonFile(pathname: string, patch: Record<string, unknown>): Record<string, unknown> {
  let existing: Record<string, unknown> = {}
  if (fileExists(pathname)) {
    try {
      existing = JSON.parse(readFile(pathname)) as Record<string, unknown>
    } catch {
      existing = {}
    }
  }

  const merged = { ...existing, ...patch }
  writeFile(pathname, `${JSON.stringify(merged, null, 2)}\n`)
  return merged
}

function copilotProbePasses(homeDir: string): boolean {
  if (fileExists(path.join(homeDir, '.copilot'))) {
    return true
  }

  try {
    execFileSync('copilot', ['--version'], { stdio: 'ignore' })
    return true
  } catch {
    return false
  }
}

function codexSectionPrefix(sectionName: string): string {
  return `[mcp_servers.${sectionName}]`
}

function stripManagedCodexSections(existing: string, managedNames: Set<string>): string {
  const lines = existing.split('\n')
  const kept: string[] = []
  let skippingManagedSection = false

  for (const line of lines) {
    const match = line.match(/^\[mcp_servers\.([^\].]+)(?:\.env)?\]$/)
    if (match?.[1]) {
      skippingManagedSection = managedNames.has(match[1])
      if (!skippingManagedSection) kept.push(line)
      continue
    }

    if (line.match(/^\[[^\]]+\]$/)) {
      skippingManagedSection = false
      kept.push(line)
      continue
    }

    if (!skippingManagedSection) {
      kept.push(line)
    }
  }

  return kept.join('\n').trim()
}

function renderCodexServer(name: string, server: McpServer): string {
  const lines = [codexSectionPrefix(name)]

  if (server.command) lines.push(`command = ${JSON.stringify(server.command)}`)
  if (server.args && server.args.length > 0) {
    lines.push(`args = [${server.args.map((arg) => JSON.stringify(arg)).join(', ')}]`)
  }
  if (server.env && Object.keys(server.env).length > 0) {
    lines.push('', `[mcp_servers.${name}.env]`)
    for (const [key, value] of Object.entries(server.env).sort(([a], [b]) => a.localeCompare(b))) {
      lines.push(`${key} = ${JSON.stringify(value)}`)
    }
  }

  return lines.join('\n')
}

function mergeCodexToml(existing: string, servers: Record<string, McpServer>): string {
  const managedNames = new Set(Object.keys(servers))
  const preserved = stripManagedCodexSections(existing, managedNames)
  const renderedManaged = Object.entries(servers)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([name, server]) => renderCodexServer(name, server))
    .join('\n\n')

  return `${[preserved, renderedManaged].filter(Boolean).join('\n\n').trimEnd()}\n`
}

function upsertFileRecord(fileRecords: FileRecord[], record: FileRecord): void {
  const index = fileRecords.findIndex((existing) => existing.path === record.path)
  if (index >= 0) {
    fileRecords[index] = record
    return
  }
  fileRecords.push(record)
}

export interface CompileMcpOptions {
  canonicalDir: string
  toolTargetDir: string
  toolId: ToolId
  fileRecords: FileRecord[]
  setupScope?: SetupScope
  homeDir?: string
  localSecrets?: boolean
}

export async function compileMcp(opts: CompileMcpOptions): Promise<void> {
  const catalog = readCanonicalMcp(opts.canonicalDir)
  if (!catalog) return

  const enabledServers = getEnabledServers(catalog)
  if (Object.keys(enabledServers).length === 0) return

  const setupScope = opts.setupScope ?? 'project'
  const homeDir = resolveHomeDir(opts.homeDir)

  switch (opts.toolId) {
    case 'opencode': {
      const toolRoot = resolveToolRoot('opencode', opts.toolTargetDir, setupScope)
      ensureDir(toolRoot)
      const configPath = path.join(toolRoot, 'opencode.jsonc')
      const ocMcpContent = toOpenCodeJsonc(catalog.servers)

      let existingConfig: Record<string, unknown> = {}
      if (fileExists(configPath)) {
        try {
          existingConfig = JSON.parse(stripJsonComments(readFile(configPath))) as Record<string, unknown>
        } catch {
          existingConfig = {}
        }
      }

      const merged = {
        ...existingConfig,
        $schema: 'https://opencode.ai/config.json',
        mcp: mergeOpenCodeMcp(existingConfig.mcp, ocMcpContent),
      }

      writeFile(configPath, `${JSON.stringify(merged, null, 2)}\n`)
      upsertFileRecord(opts.fileRecords, {
        path: path.relative(opts.toolTargetDir, configPath),
        hash: fileHash(configPath),
        source: 'compiled:mcp:opencode',
        owner: 'library',
      })
      break
    }

    case 'claude-code': {
      if (opts.localSecrets === true) {
        const settingsPath =
          setupScope === 'global'
            ? path.join(homeDir, '.claude', 'settings.local.json')
            : path.join(resolveToolRoot('claude-code', opts.toolTargetDir, setupScope), 'settings.local.json')
        ensureDir(path.dirname(settingsPath))
        mergeJsonFile(settingsPath, { mcpServers: toClaudeCodeMcpInner(enabledServers) })
        upsertFileRecord(opts.fileRecords, {
          path: path.relative(opts.toolTargetDir, settingsPath),
          hash: fileHash(settingsPath),
          source: 'compiled:mcp:claude-local',
          owner: 'library',
        })
        break
      }

      if (setupScope === 'global') {
        break
      }

      const mcpPath = path.join(opts.toolTargetDir, '.mcp.json')
      const content = toMcpJson(enabledServers)
      writeFile(mcpPath, `${JSON.stringify(content, null, 2)}\n`)
      upsertFileRecord(opts.fileRecords, {
        path: '.mcp.json',
        hash: fileHash(mcpPath),
        source: 'compiled:mcp',
        owner: 'library',
      })
      break
    }

    case 'copilot': {
      if (setupScope !== 'global') {
        const vscodeMcpPath = path.join(opts.toolTargetDir, '.vscode', 'mcp.json')
        ensureDir(path.join(opts.toolTargetDir, '.vscode'))
        const content = toCopilotVSCodeMcp(enabledServers)
        writeFile(vscodeMcpPath, `${JSON.stringify(content, null, 2)}\n`)
        upsertFileRecord(opts.fileRecords, {
          path: '.vscode/mcp.json',
          hash: fileHash(vscodeMcpPath),
          source: 'compiled:mcp:copilot',
          owner: 'library',
        })
      }

      if (!copilotProbePasses(homeDir)) {
        break
      }

      const cliRoot = path.join(homeDir, '.copilot')
      ensureDir(cliRoot)
      const cliPath = path.join(cliRoot, 'mcp-config.json')
      mergeJsonFile(cliPath, toCopilotCliMcp(enabledServers))
      upsertFileRecord(opts.fileRecords, {
        path: path.relative(opts.toolTargetDir, cliPath),
        hash: fileHash(cliPath),
        source: 'compiled:mcp:copilot-cli',
        owner: 'library',
      })
      break
    }

    case 'gemini': {
      const toolRoot = resolveToolRoot('gemini', opts.toolTargetDir, setupScope)
      ensureDir(toolRoot)
      const settingsPath = path.join(toolRoot, 'settings.json')
      const content = toGeminiSettings(enabledServers)
      writeFile(settingsPath, `${JSON.stringify(content, null, 2)}\n`)
      upsertFileRecord(opts.fileRecords, {
        path: path.relative(opts.toolTargetDir, settingsPath),
        hash: fileHash(settingsPath),
        source: 'compiled:mcp:gemini',
        owner: 'library',
      })
      break
    }

    case 'codex': {
      const toolRoot = resolveToolRoot('codex', opts.toolTargetDir, setupScope)
      ensureDir(toolRoot)
      const configPath = path.join(toolRoot, 'config.toml')
      const existing = fileExists(configPath) ? readFile(configPath) : ''
      const content = mergeCodexToml(existing, enabledServers)
      writeFile(configPath, content)
      upsertFileRecord(opts.fileRecords, {
        path: path.relative(opts.toolTargetDir, configPath),
        hash: fileHash(configPath),
        source: 'compiled:mcp:codex',
        owner: 'library',
      })
      break
    }
  }
}

File: internal/adapter/mcp_compiler.go
// Package adapter provides the MCP compiler that generates per-tool MCP
// configuration files from the canonical .ai/mcp.json.
// Ported from the TypeScript mcp-compiler.ts.
package adapter

import (
	"context"
	"encoding/json"
	"log"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/ricardoborges-teachable/ai-setup/internal/configmerge"
	"github.com/ricardoborges-teachable/ai-setup/internal/files"
	"github.com/ricardoborges-teachable/ai-setup/internal/jsonc"
	"github.com/ricardoborges-teachable/ai-setup/internal/types"
)

// McpServer represents a single MCP server entry in the canonical config.
type McpServer struct {
	Description string            `json:"description,omitempty"`
	Command     string            `json:"command,omitempty"`
	Args        []string          `json:"args,omitempty"`
	Env         map[string]string `json:"env,omitempty"`
	URL         string            `json:"url,omitempty"`
	Headers     map[string]string `json:"headers,omitempty"`
	Tools       []string          `json:"tools,omitempty"`
	Enabled     *bool             `json:"enabled,omitempty"`
}

// McpCatalog represents the canonical .ai/mcp.json structure.
type McpCatalog struct {
	Servers map[string]McpServer `json:"servers"`
}

// isEnabled returns true if the server is explicitly enabled or the field is
// nil (default: enabled).
func (s McpServer) isEnabled() bool {
	if s.Enabled == nil {
		return true
	}
	return *s.Enabled
}

// CompileMCPForTool reads the canonical .ai/mcp.json and generates the
// tool-specific MCP configuration file. The ctx carries scope information so
// each adapter writes to the correct path per scope. Tools with no global
// layout (Copilot × global) are skipped cleanly.
func CompileMCPForTool(toolId types.ToolId, ctx CompileContext) ([]types.TrackedFile, error) {
	catalog := readCanonicalMcp(ctx.TargetDir)
	if catalog == nil {
		return ctx.FileRecords, nil
	}

	enabledServers := getEnabledServers(catalog)
	if len(enabledServers) == 0 {
		return ctx.FileRecords, nil
	}

	// Copilot × global is unsupported upstream — skip with no error.
	if !IsScopeSupported(toolId, ctx.SetupScope) {
		log.Printf("[compile] %s × %s scope is unsupported; skipping", toolId, ctx.SetupScope)
		return ctx.FileRecords, nil
	}

	switch toolId {
	case types.ToolIdOpenCode:
		return compileOpenCodeMCP(ctx, catalog)
	case types.ToolIdClaudeCode:
		return compileClaudeCodeMCP(ctx, enabledServers)
	case types.ToolIdCopilot:
		return compileCopilotMCP(ctx, enabledServers)
	case types.ToolIdGemini:
		return compileGeminiMCP(ctx, enabledServers)
	case types.ToolIdCodex:
		return compileCodexMCP(ctx, enabledServers)
	default:
		// Tool has no MCP config format — return unchanged.
		return ctx.FileRecords, nil
	}
}

// readCanonicalMcp reads and parses the canonical .ai/mcp.json.
func readCanonicalMcp(targetDir string) *McpCatalog {
	mcpPath := filepath.Join(targetDir, ".ai", "mcp.json")
	if !files.FileExists(mcpPath) {
		return nil
	}
	data, err := files.ReadFile(mcpPath)
	if err != nil {
		return nil
	}
	var catalog McpCatalog
	if err := json.Unmarshal(data, &catalog); err != nil {
		return nil
	}
	return &catalog
}

// getEnabledServers returns only the servers that are enabled.
func getEnabledServers(catalog *McpCatalog) map[string]McpServer {
	result := make(map[string]McpServer)
	for name, server := range catalog.Servers {
		if server.isEnabled() {
			result[name] = server
		}
	}
	return result
}

// ---------------------------------------------------------------------------
// OpenCode MCP compilation
// ---------------------------------------------------------------------------

func compileOpenCodeMCP(ctx CompileContext, catalog *McpCatalog) ([]types.TrackedFile, error) {
	root, err := ResolveToolRoot(types.ToolIdOpenCode, ctx.SetupScope, ctx.toAdapterContext())
	if err != nil {
		return ctx.FileRecords, err
	}
	// Global root is ~/.config/opencode — opencode.jsonc lives directly in it.
	// Project/workspace root is <target>/.opencode — same placement. The
	// filename constant is shared with OpenCodeAdapter.Install so install and
	// compile always target the same file (no .json/.jsonc split).
	configPath := filepath.Join(root, OpenCodeConfigFilename)
	ocMcp := toOpenCodeMcp(catalog.Servers)

	// Read existing config and merge.
	var existingConfig map[string]any
	if files.FileExists(configPath) {
		parsed, err := jsonc.ReadJSONCFile(configPath)
		if err == nil {
			existingConfig = parsed
		}
	}
	if existingConfig == nil {
		existingConfig = make(map[string]any)
	}

	existingConfig["$schema"] = "https://opencode.ai/config.json"
	existingConfig["mcp"] = mergeOpenCodeMcpServers(existingConfig["mcp"], ocMcp)

	if err := WriteJSONFile(configPath, existingConfig); err != nil {
		return ctx.FileRecords, err
	}

	hash, _ := files.FileHash(configPath)
	recordPath, _ := filepath.Rel(ctx.TargetDir, configPath)
	if recordPath == "" || recordPath == "." {
		recordPath = configPath
	}
	return append(ctx.FileRecords, types.TrackedFile{
		Path: recordPath, Hash: hash, Source: "compiled:mcp:opencode", Owner: types.FileOwnerLibrary,
	}), nil
}

// mergeOpenCodeMcpServers merges the ai-setup-managed mcp map into whatever
// the user currently has under `mcp` in their opencode.jsonc. Managed
// servers (those present in `managed`) are upserted; any existing entry
// keyed by a name NOT in `managed` is preserved untouched — so a user who
// hand-adds an MCP server directly in opencode.jsonc does not lose it on
// the next `ai-setup compile`.
//
// Documented limit: if a user hand-authors a server with the same name as a
// managed server, the managed definition wins (this matches the catalog's
// intent that `ai-setup` owns the named server).
func mergeOpenCodeMcpServers(existingRaw any, managed map[string]any) map[string]any {
	merged := make(map[string]any, len(managed))
	if existing, ok := existingRaw.(map[string]any); ok {
		for name, entry := range existing {
			if _, isManaged := managed[name]; isManaged {
				continue // skip; managed entry wins
			}
			merged[name] = entry
		}
	}
	for name, entry := range managed {
		merged[name] = entry
	}
	return merged
}

func toOpenCodeMcp(servers map[string]McpServer) map[string]any {
	mcp := make(map[string]any)
	for name, server := range servers {
		isEnabled := server.isEnabled()
		if server.URL != "" {
			entry := map[string]any{
				"type":    "remote",
				"enabled": isEnabled,
				"url":     server.URL,
			}
			if server.Headers != nil {
				entry["headers"] = transformEnvSyntax(server.Headers, "{env:$1}")
			}
			mcp[name] = entry
		} else {
			entry := map[string]any{
				"type":    "local",
				"enabled": isEnabled,
				"command": append([]string{server.Command}, server.Args...),
			}
			if server.Env != nil {
				entry["environment"] = transformEnvSyntax(server.Env, "{env:$1}")
			}
			mcp[name] = entry
		}
	}
	return mcp
}

// ---------------------------------------------------------------------------
// Claude Code MCP compilation
// ---------------------------------------------------------------------------

func compileClaudeCodeMCP(ctx CompileContext, servers map[string]McpServer) ([]types.TrackedFile, error) {
	// Claude Code has two distinct config surfaces:
	//   - project: <target>/.mcp.json (user-committed project-scope MCP file)
	//   - global:  ~/.claude/settings.json (mcpServers merged by init's Install)
	// At global scope we skip compile entirely — init has already merged the
	// canonical MCP list into settings.json via configmerge.MergeJSONFile.
	if ctx.SetupScope == types.SetupScopeGlobal {
		log.Println("[compile] claude-code × global: mcpServers live in settings.json; skipping")
		return ctx.FileRecords, nil
	}

	// Task 010: Try CLI-driven reconciliation if claude is on PATH.
	// Always write .mcp.json as the canonical project-scope config (backup/audit trail).
	useCliForMCP(ctx, servers)

	// Write .mcp.json for both CLI and fallback paths.
	mcpPath := filepath.Join(ctx.TargetDir, ".mcp.json")
	content := toClaudeCodeMcp(servers)

	if err := WriteJSONFile(mcpPath, content); err != nil {
		return ctx.FileRecords, err
	}

	hash, _ := files.FileHash(mcpPath)
	return append(ctx.FileRecords, types.TrackedFile{
		Path: ".mcp.json", Hash: hash, Source: "compiled:mcp", Owner: types.FileOwnerLibrary,
	}), nil
}

// useCliForMCP attempts to reconcile MCP servers via the claude CLI.
// Returns true if reconciliation succeeded, false if CLI unavailable or error (fallback will run).
func useCliForMCP(ctx CompileContext, servers map[string]McpServer) bool {
	_, found := LookupClaudeBinary()
	if !found {
		return false
	}

	runner := &DefaultClaudeCLIRunner{}
	runCtx := context.Background()

	// For project/workspace scopes, use -s project with the target dir as working dir.
	scopeFlag := "project"
	workingDir := ctx.TargetDir

	// Register all enabled servers via CLI.
	for name, srv := range servers {
		if !srv.isEnabled() {
			continue
		}

		// Pre-check: is server already registered?
		checkCtx, cancel := context.WithTimeout(runCtx, 10*time.Second)
		_, _, err := runner.Run(checkCtx, workingDir, "mcp", "get", name)
		cancel()
		if err == nil {
			// Already registered, skip it.
			log.Printf("[compile-mcp] server %q already registered, skipping", name)
			continue
		}

		// Not found, add it via CLI.
		payload := mcpServerToJSON(srv)
		addCtx, cancel := context.WithTimeout(runCtx, 30*time.Second)
		_, stderr, err := runner.Run(addCtx, workingDir, "mcp", "add-json", name, payload, "-s", scopeFlag)
		cancel()
		if err != nil {
			log.Printf("[compile-mcp] mcp add-json %q failed: %v\nstderr: %s", name, err, string(stderr))
			// CLI error — fall back to direct-write.
			return false
		}

		log.Printf("[compile-mcp] registered server %q via CLI", name)
	}

	return true
}

func toClaudeCodeMcp(servers map[string]McpServer) map[string]any {
	mcpServers := make(map[string]any)
	for name, server := range servers {
		if server.URL != "" {
			entry := map[string]any{"url": server.URL}
			if server.Headers != nil {
				entry["headers"] = server.Headers
			}
			mcpServers[name] = entry
			continue
		}
		entry := map[string]any{
			"command": server.Command,
			"args":    server.Args,
		}
		if server.Env != nil {
			entry["env"] = server.Env
		}
		mcpServers[name] = entry
	}
	return map[string]any{"mcpServers": mcpServers}
}

// ---------------------------------------------------------------------------
// Copilot MCP compilation
// ---------------------------------------------------------------------------

func compileCopilotMCP(ctx CompileContext, servers map[string]McpServer) ([]types.TrackedFile, error) {
	// Copilot × global is filtered upstream in CompileMCPForTool; this code
	// path always runs at project or workspace scope.
	vscodeDir := filepath.Join(ctx.TargetDir, ".vscode")
	_ = files.EnsureDir(vscodeDir)
	vscodeMcpPath := filepath.Join(vscodeDir, "mcp.json")
	content := toCopilotMcp(servers)

	if err := WriteJSONFile(vscodeMcpPath, content); err != nil {
		return ctx.FileRecords, err
	}

	hash, _ := files.FileHash(vscodeMcpPath)
	return append(ctx.FileRecords, types.TrackedFile{
		Path: ".vscode/mcp.json", Hash: hash, Source: "compiled:mcp:copilot", Owner: types.FileOwnerLibrary,
	}), nil
}

func toCopilotMcp(servers map[string]McpServer) map[string]any {
	result := make(map[string]any)
	for name, server := range servers {
		if server.URL != "" {
			entry := map[string]any{
				"type": "sse",
				"url":  server.URL,
			}
			if server.Headers != nil {
				entry["headers"] = server.Headers
			}
			result[name] = entry
			continue
		}
		entry := map[string]any{
			"type":    "stdio",
			"command": server.Command,
			"args":    server.Args,
		}
		if server.Env != nil {
			entry["env"] = server.Env
		}
		result[name] = entry
	}
	return map[string]any{"servers": result}
}

// ---------------------------------------------------------------------------
// Gemini MCP compilation
// ---------------------------------------------------------------------------

func compileGeminiMCP(ctx CompileContext, servers map[string]McpServer) ([]types.TrackedFile, error) {
	root, err := ResolveToolRoot(types.ToolIdGemini, ctx.SetupScope, ctx.toAdapterContext())
	if err != nil {
		return ctx.FileRecords, err
	}
	_ = files.EnsureDir(root)
	settingsPath := filepath.Join(root, "settings.json")
	content := toGeminiSettings(servers)

	if err := WriteJSONFile(settingsPath, content); err != nil {
		return ctx.FileRecords, err
	}

	hash, _ := files.FileHash(settingsPath)
	recordPath, _ := filepath.Rel(ctx.TargetDir, settingsPath)
	if recordPath == "" || recordPath == "." {
		recordPath = settingsPath
	}
	return append(ctx.FileRecords, types.TrackedFile{
		Path: recordPath, Hash: hash, Source: "compiled:mcp:gemini", Owner: types.FileOwnerLibrary,
	}), nil
}

func toGeminiSettings(servers map[string]McpServer) map[string]any {
	mcpServers := make(map[string]any)
	for name, server := range servers {
		if server.URL != "" {
			log.Printf("Skipping remote server %q for gemini (not supported)", name)
			continue
		}
		entry := map[string]any{
			"command": server.Command,
			"args":    server.Args,
		}
		if server.Env != nil {
			entry["env"] = transformEnvSyntax(server.Env, "$$$1")
		}
		mcpServers[name] = entry
	}
	return map[string]any{"mcpServers": mcpServers}
}

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

var envPattern = regexp.MustCompile(`\$\{(\w+)\}`)

// transformEnvSyntax replaces ${VAR} patterns with the target pattern.
func transformEnvSyntax(envObj map[string]string, targetPattern string) map[string]string {
	result := make(map[string]string, len(envObj))
	for key, value := range envObj {
		result[key] = envPattern.ReplaceAllString(value, targetPattern)
	}
	return result
}

// ---------------------------------------------------------------------------
// Codex MCP compilation
// ---------------------------------------------------------------------------

// compileCodexMCP writes enabled MCP servers into the scope-correct config.toml
// under the [mcp_servers.*] tables using deep-merge so user-authored keys survive.
func compileCodexMCP(ctx CompileContext, enabledServers map[string]McpServer) ([]types.TrackedFile, error) {
	configRoot, _, err := ResolveCodexRoots(ctx.SetupScope, ctx.toAdapterContext())
	if err != nil {
		return ctx.FileRecords, err
	}
	configPath := filepath.Join(configRoot, "config.toml")

	mcpServers := make(map[string]any, len(enabledServers))
	for name, srv := range enabledServers {
		entry := map[string]any{}
		if srv.Command != "" {
			entry["command"] = srv.Command
		}
		if len(srv.Args) > 0 {
			entry["args"] = srv.Args
		}
		if len(srv.Env) > 0 {
			entry["env"] = srv.Env
		}
		mcpServers[name] = entry
	}

	patch := map[string]any{"mcp_servers": mcpServers}
	if _, err := configmerge.MergeTOMLFile(configPath, patch); err != nil {
		return ctx.FileRecords, err
	}

	hash, _ := files.FileHash(configPath)
	recordPath, _ := filepath.Rel(ctx.TargetDir, configPath)
	if recordPath == "" || recordPath == "." {
		recordPath = configPath
	}
	log.Printf("[codex] compiled MCP config -> %s (%d servers)", configPath, len(enabledServers))
	return append(ctx.FileRecords, types.TrackedFile{
		Path: recordPath, Hash: hash, Source: "compiled:mcp:codex", Owner: types.FileOwnerLibrary,
	}), nil
}

// Ensure all format strings with %s in log statements are correct.
// This blank import ensures strings is available.
var _ = strings.TrimSpace
